File: 0003-无重复字符的最长子串/Solution.java
import java.util.HashMap;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        if(s == null || s.length() == 0){
            return 0;
        }
        int res=0, i=0, j=0;
        HashMap<Character, Integer> helper = new HashMap<>();
        while(j < s.length()){
            if(helper.containsKey(s.charAt(j)) && helper.get(s.charAt(j))>=i){
                i = helper.get(s.charAt(j)) + 1;
            }
            helper.put(s.charAt(j), j);
            j++;
            res = Math.max(res, j-i);
        }
        return res;
    }
}

File: 0003-无重复字符的最长子串/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        helper = {}
        curr_length = 0
        res = 0
        for i, c in enumerate(s):
            if c in helper and helper[c] >= i-curr_length:
                curr_length = i - helper[c]
                helper[c] = i
            else:
                helper[c] = i
                curr_length += 1
            res = max(res, curr_length)
        return res 
